### PR TITLE
Fix when primary key is not "int" or "long". Fixes #380

### LINKED ISSUE #380: Constraint failed on primary key when using bulk insert
Exception:
```
   at Microsoft.Data.Sqlite.SqliteException.ThrowExceptionForRC(Int32 rc, sqlite3 db)
   at Microsoft.Data.Sqlite.SqliteDataReader.NextResult()
   at Microsoft.Data.Sqlite.SqliteCommand.ExecuteReader(CommandBehavior behavior)
   at Microsoft.Data.Sqlite.SqliteCommand.ExecuteReader()
   at Microsoft.Data.Sqlite.SqliteCommand.ExecuteNonQuery()
   at EFCore.BulkExtensions.SqlBulkOperation.Merge[T](DbContext context, Type type, IList`1 entities, TableInfo tableInfo, OperationType operationType, Action`1 progress)
   at EFCore.BulkExtensions.SqlBulkOperation.Merge[T](DbContext context, IList`1 entities, TableInfo tableInfo, OperationType operationType, Action`1 progress)
   at EFCore.BulkExtensions.DbContextBulkTransaction.Execute[T](DbContext context, IList`1 entities, OperationType operationType, BulkConfig bulkConfig, Action`1 progress)
   at EFCore.BulkExtensions.DbContextBulkExtensions.BulkInsert[T](DbContext context, IList`1 entities, Action`1 bulkAction, Action`1 progress)
   at RomRepoMgr.Core.Workers.DatImporter.Import() in /home/claunia/Development/RomRepoMgr/RomRepoMgr.Core/Workers/DatImporter.cs:line 187
```

Model:
```csharp
    public class Machine : BaseModel<ulong>
    {
        [Required]
        public string Name { get; set; }
        [Required]
        public virtual RomSet RomSet { get;                        set; }
        public virtual ICollection<FileByMachine>  Files    { get; set; }
        public virtual ICollection<DiskByMachine>  Disks    { get; set; }
        public virtual ICollection<MediaByMachine> Medias   { get; set; }
        public         long                        RomSetId { get; set; }
    }
```

Base model:
```csharp
    public abstract class BaseModel<TKey>
    {
        public TKey Id { get; set; }

        public DateTime CreatedOn { get; set; }
        public DateTime UpdatedOn { get; set; }
    }
```

Context:
```csharp
        public DbSet<Machine>        Machines         { get; set; }

            modelBuilder.Entity<Machine>(entity =>
            {
                entity.HasIndex(e => e.Name);

                entity.HasOne(e => e.RomSet).WithMany(e => e.Machines).OnDelete(DeleteBehavior.Cascade);
            });
```

Snapshot:
```csharp
modelBuilder.Entity("RomRepoMgr.Database.Models.Machine", b =>
            {
                b.Property<ulong>("Id").ValueGeneratedOnAdd().HasColumnType("INTEGER");

                b.Property<DateTime>("CreatedOn").HasColumnType("TEXT");

                b.Property<string>("Name").IsRequired().HasColumnType("TEXT");

                b.Property<long>("RomSetId").HasColumnType("INTEGER");

                b.Property<DateTime>("UpdatedOn").HasColumnType("TEXT");

                b.HasKey("Id");

                b.HasIndex("Name");

                b.HasIndex("RomSetId");

                b.ToTable("Machines");
            });
```

Insertion code:
```csharp
                Dictionary<string, Machine> machines = new Dictionary<string, Machine>();

                foreach(string name in machineNames)
                {
                    var machine = new Machine
                    {
                        Name      = name,
                        RomSetId  = romSet.Id, // 1
                        CreatedOn = DateTime.UtcNow,
                        UpdatedOn = DateTime.UtcNow
                    };

                    machines[name] = machine;
                }

                ctx.BulkInsert(machines.Values.ToList(), b => b.SetOutputIdentity=true);
```

### PR DESCRIPTION
**Pull request by claunia**

### CODE REVIEW

File: EFCore.BulkExtensions/SqlBulkOperation.cs [hunk 1]
```diff
@@ -502,21 +554,71 @@ public static async Task MergeAsync(DbContext context, Type type, IList<object>
                             command.CommandText = SqlQueryBuilderSqlite.SelectLastInsertRowId();
                             long lastRowIdScalar = (long)await command.ExecuteScalarAsync(cancellationToken).ConfigureAwait(false);
                             var identityPropertyInteger = false;
+                            var identityPropertyUnsigned = false;
+                            var identityPropertyByte = false;
+                            var identityPropertyShort = false;
                             var accessor = TypeAccessor.Create(type, true);
                             string identityPropertyName = tableInfo.PropertyColumnNamesDict.SingleOrDefault(a => a.Value == tableInfo.IdentityColumnName).Key;
-                            if (accessor.GetMembers().FirstOrDefault(x => x.Name == identityPropertyName)?.Type == typeof(int))
+                            if (accessor.GetMembers().FirstOrDefault(x => x.Name == identityPropertyName)?.Type == typeof(ulong))
```
Comment (gandhis1) [inline EFCore.BulkExtensions/SqlBulkOperation.cs:562]: You have `ulong`, do you need to check `long` as well?

**Review by gandhis1 — COMMENTED** [review]
I have limited familiarity with this code, but I am wondering if there's any way to do this that doesn't need to be so aware of the individual data types and have a whole logic branch for each.

File: EFCore.BulkExtensions/SqlBulkOperation.cs (hunk 1 above)
Comment (claunia) [reply EFCore.BulkExtensions/SqlBulkOperation.cs:562]: The problem is down the line when the Id property is set using reflection, you need to cast it to the exact type the .NET property is, or you will get an exception.

I couldn't find a way to do that in less lines of code that was not directly an order of magnitude slower.

**Comment by borisdj** [comment]
Could you resolve the conflicts from the merge.

File: EFCore.BulkExtensions/SqlBulkOperation.cs (hunk 1 above)
Comment (Numpsy) [reply EFCore.BulkExtensions/SqlBulkOperation.cs:562]: Does it need to duplicate the logic between Merge and MergeAsync, or could it be shared?

**Comment by borisdj** [comment]
Since conflict are not resolved, I have added this fix manually, NuGet 3.2.6
Thx for contrib.

**Comment by claunia** [comment]
Sorry I did not have time to look at it up. Happy to see you were able to fix it manually!

**Pull request closed without merging**